pkg/ctl/install/install.go

```diff
@@ -0,0 +1,435 @@
+	}
+
+	// Confirm there is something to commit, otherwise move on
+	diffCtx, diffCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer diffCtxCancel()
+	if err := runGitCmd(diffCtx, cloneDir, "diff", "--cached", "--quiet", "--", fi.opts.gitFluxPath); err == nil {
+		logger.Info("Nothing to commit (the repository contained identical manifests), moving on")
+		return nil
+	} else if  _, ok := err.(*exec.ExitError); !ok{
+		return err
+	}
+
+	// Commit
+	commitCtx, commitCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer commitCtxCancel()
+	args := []string{"commit",
+		"-m", "Add Initial Flux configuration",
+		fmt.Sprintf("--author=%s <%s>", fi.opts.templateParams.GitUser, fi.opts.templateParams.GitEmail),
+	}
+	if err := runGitCmd(commitCtx, cloneDir, args...); err != nil {
+		return err
+	}
+
+	// Push
+	pushCtx, pushCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer pushCtxCancel()
+	return runGitCmd(pushCtx, cloneDir, "push")
+}
+
+func (fi *fluxInstaller) applyManifestsInDir(ctx context.Context, dir string) error {
+	// NOTE: invoking kubectl is not ideal, but it's the less hairy way to do it until
+	//       server-side apply is available ( https://github.com/kubernetes/enhancements/issues/555 )
+	applyCtx, applyCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer applyCtxCancel()
+	kubeconfigFile, err := ioutil.TempFile(os.TempDir(), "eksctl-flux-install-kubeconfig")
+	if err != nil {
+		return fmt.Errorf("cannot find temporary kubeconfig file: %s", err)
+	}
+	kubeconfigPath := kubeconfigFile.Name()
+	kubeconfigFile.Close()
+	// defer os.Remove(kubeconfigFile.Name())
+	if _, err := kubeconfig.Write(kubeconfigPath, *fi.k8sConfig, true); err != nil {
+		return fmt.Errorf("cannot write to temporary kubeconfig file (%s): %s", kubeconfigFile.Name(), err)
+	}
+
+	kubectlCmd := exec.CommandContext(applyCtx, "kubectl", "apply", "-f", dir)
+	kubectlCmd.Env = []string{"KUBECONFIG="+kubeconfigFile.Name()}
+	kubectlCmd.Stdout = os.Stdout
+	kubectlCmd.Stderr = os.Stderr
+	return kubectlCmd.Run()
+}
+
+func runGitCmd(ctx context.Context, dir string, args ...string) error {
+	gitCmd := exec.CommandContext(ctx, "git", args...)
+	gitCmd.Stdout = os.Stdout
+	gitCmd.Stderr = os.Stderr
+	gitCmd.Dir = dir
+	return gitCmd.Run()
+}
+
+func getFluxManifests(params install.TemplateParameters, cs *kubernetes.Clientset) (map[string][]byte, error) {
+	params.AdditionalFluxArgs = []string{"--sync-garbage-collection", "--manifest-generation"}
+	manifests, err := install.FillInTemplates(params)
+	if err != nil {
+		return nil, err
+	}
+	// Verify if the namespace to install Flux already exists and otherwise add a manifest for it
+	_, err = cs.CoreV1().Namespaces().Get(params.Namespace, metav1.GetOptions{})
+	if err == nil {
+		return manifests, nil
+	}
+	if !k8serrors.IsNotFound(err) {
+		return nil, fmt.Errorf("cannot check whether namespace %s exists: %s", params.Namespace, err)
+	}
+	nsTemplate := `---
+apiVersion: v1
```

Comment:
Please don't do this. If you must use strings, turn it into JSON and use appropriate number of spaces/tabs, or better - define `CreateNamespace` function in `pkg/kubernetes` and use a struct.

Reply:
I think that's overkill, considering you need to serialize it to disk and it needs a --- prefix to work

Reply:
FWIW, I wanted to try using a `struct`...
```go
package kubernetes

import (
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// Namespace creates a corev1.Namespace object with the standard labels set
// using the provided name.
func Namespace(name string) *corev1.Namespace {
	return &corev1.Namespace{
		ObjectMeta: metav1.ObjectMeta{
			Name:   name,
			Labels: map[string]string{"name": name},
		},
	}
}
```
... and serialising it, however, empty/null objects are getting serialised, which is rather annoying:
```go
package kubernetes_test

import (
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
	. "github.com/weaveworks/eksctl/pkg/kubernetes"
	"k8s.io/apimachinery/pkg/runtime/serializer/json"
	"k8s.io/client-go/kubernetes/scheme"
	"sigs.k8s.io/yaml"
)

const actualYAML = `apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: null
  labels:
    name: flux
  name: flux
spec: {}
status: {}
`

func TestSerialiseViaSigsK8SIOYAML(t *testing.T) {
	ns := Namespace("flux")
	yamlBytes, err := yaml.Marshal(ns)
	assert.NoError(t, err)
	assert.Equal(t, actualYAML, string(yamlBytes))
}

func TestSerialiseUsingInnerYAMLSerializer(t *testing.T) {
	ns := Namespace("flux")
	s := json.NewYAMLSerializer(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme)
	var builder strings.Builder
	err := s.Encode(ns, &builder)
	assert.NoError(t, err)
	assert.Equal(t, actualYAML, builder.String())
}
```
Also, for better or worse, fields are alphabetically sorted. 
Unless anyone has a better suggestion, I'll use a template `string` instead for now.

Reply:
@marccarre I think you get the empty fields if you use `go-yaml/yaml` but try using our yaml encoder "github.com/ghodss/yaml". That one omits the empty fields I think.

Reply:
@martina-if, thanks for your reply! However, I still get empty fields if I replace `"sigs.k8s.io/yaml"` with `"github.com/ghodss/yaml`:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: null
  labels:
    name: flux
  name: flux
spec: {}
status: {}
```
I believe this is because the `struct` has other `struct` as fields, as opposed to pointers to `struct`s. @errordeveloper shared a clean-up function he wrote (https://github.com/errordeveloper/kubegen/blob/master/pkg/util/cleanup.go), but I think it would be pushing a bit too far for such a simple application. Something to consider if we end up doing more of this though.

Reply:
Relevant changes: https://github.com/weaveworks/eksctl/pull/1064/commits/b55c3f42ad784ad28a1396f4b2fc7930eeb115f8 and https://github.com/weaveworks/eksctl/pull/1064/commits/9d618f9185b980353a3236726085cb2eab3ba272